libsupport/include/katana/Iterators.h

```diff
@@ -0,0 +1,187 @@
+#ifndef KATANA_LIBSUPPORT_KATANA_ITERATORS_H_
+#define KATANA_LIBSUPPORT_KATANA_ITERATORS_H_
+
+#include <algorithm>
+#include <utility>
+#include <vector>
+
+#include <boost/iterator/iterator_facade.hpp>
+
+namespace katana {
+
+//TODO(amber): Move other iterators from libgalois to libsupport
+
+/**
+ * A Zip Iterator allows iterating over several containers simultaneously.
+ * A canonical example is sorting two containers together. A Zip iterator maintains
+ * a tuple of participating iterators and advances all of them together.
+ * Dereferencing yields a tuple of references for l-values and a tuple of values
+ * for r-values. See  code below:
+ * for example usage:
+
+ * NOTE: It is best to define comparator functions as template functions (or
+ * lambdas that take parameters as 'auto'), e.g., for std::sort(),
+ * because either left or right parameter to the comparator can be a
+ * tuple of references or tuple of values depending on the calling context.
+
+int main() {
+
+  std::vector<int> vecA{ 5, 3, 4, 2, 1};
+  std::vector<char> vecB{ 'e', 'c', 'd', 'b', 'a' };
+
+  auto beg = katana::make_zip_iterator(vecA.begin(), vecB.begin());
+
+  auto end = katana::make_zip_iterator(vecA.end(), vecB.end());
+
+  std::cout << "[";
+  for (auto i = beg; i != end; ++i) {
+    std::cout << "(" << std::get<0>(*i) << "," << std::get<1>(*i) << "), ";
+  }
+  std::cout << "]" << std::endl;
+
+  std::sort(beg, end,
+      [&] (const auto& tup1, const auto& tup2) {
+        return std::get<0>(tup1) < std::get<0>(tup2);
+      });
+
+
+  std::cout << "[";
+  for (size_t i = 0; i < vecA.size(); ++i) {
+    std::cout << "(" << vecA[i] << "," << vecB[i] << "), ";
+  }
+  std::cout << "]" << std::endl;
+
+  return 0;
+}
+
+*/
+
+/**
+ * This is a thin wrapper around std::tuple<Args&...> . It is needed to
```

Comment:
Can you explain why this is different than https://www.boost.org/doc/libs/1_41_0/libs/iterator/doc/zip_iterator.html ?

Reply:
I was seeing errors due to ill-defined `swap` originating from `libstdc++` code. I can reproduce a small example for you. 

The other issue is that `boost::tuple` has a different `get<>()` syntax from `std::tuple` and I chose to prefer `std::tuple` syntax. It's simpler. boost syntax requires things like `tup.template get<0>()`.

Reply:
fyi: I started with `boost::zip_iterator` before going down this rabbit hole and ruining my Saturday (last weekend). I suspect there might be a way to fix `boost::zip_iterator` but then I don't like the `boost::tuple` syntax.

Reply:
Is this the same as `boost::combine`?

Reply:
Some similarities. We'll have to test it against `std::sort()` to see if it really works. `std::sort()` implementation is a good test because it performs `swap` and move/copy assignment in its code.